google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/OpenCensusUtil.java

```diff
@@ -0,0 +1,193 @@
+    if (traceOpts.isSampled()) {
+      return "";
+    }
+    return "traceid=" + traceId.toLowerBase16()
+        + "&spanid=" + spanId.toLowerBase16()
+        + "&traceopt=" + (traceOpts.isSampled() ? "t&" : "f&");
+  }
+
+  private static String encodeTagContext(TagContext tags) {
+    try {
+      byte[] encodedTags = serializer.toByteArray(tags);
+      return Base64.getEncoder().encodeToString(encodedTags);
+    }
+    catch (TagContextSerializationException exn) {
+      logger.log(Level.INFO, "OpenCensus: Tag Context Serialization Exception: " + exn);
+      return "";
+    }
+  }
+
+  private static Scope createScopedTagContext(String encodedTags) {
+    try {
+      TagContext tags = serializer.fromByteArray(Base64.getDecoder().decode(encodedTags));
+      return tagger.withTagContext(tags);
+    } catch (TagContextDeserializationException exn) {
+      logger.log(Level.INFO, "OpenCensus: Tag Context Deserialization Exception: " + exn);
+      return tagger.withTagContext(tagger.getCurrentTagContext());
+    }
+  }
+
+  @MustBeClosed
+  private static Scope createScopedSpan(String name) {
+    return tracer
+        .spanBuilderWithExplicitParent(name, tracer.getCurrentSpan())
+        .setRecordEvents(true)
+        .startScopedSpan();
+  }
+
+  private static void addParentLink(String encodedParentSpanContext) {
+    tracer.getCurrentSpan().addLink(Link.fromSpanContext(
+        createSpanContext(encodedParentSpanContext), Link.Type.PARENT_LINKED_SPAN));
+  }
+
+  private static SpanContext createSpanContext(String encodedSpanContext) {
+    String traceId = getTraceId(encodedSpanContext);
+    String spanId = getSpanId(encodedSpanContext);
+    String traceOpt = getTraceOpt(encodedSpanContext);
+    return SpanContext.create(
+        TraceId.fromLowerBase16(traceId),
+        SpanId.fromLowerBase16(spanId),
+        traceOpt.equals("t") ? SAMPLED : TraceOptions.DEFAULT);
+  }
+
+  private static String getTraceId(String encodedSpan) {
+    return lookupKey("traceid=", encodedSpan);
+  }
+
+  private static String getSpanId(String encodedSpan) {
+    return lookupKey("spanid=", encodedSpan);
+  }
+
+  private static String getTraceOpt(String encodedSpan) {
+    return lookupKey("traceopt=", encodedSpan);
+  }
+
+  // encodedSpan = (key=value&)*
+  private static String lookupKey(String key, String encodedSpan) {
+    int start = encodedSpan.indexOf(key, 0);
+    if (start == -1) {
+      return "";
+    }
+    start += key.length();
+    int end = encodedSpan.indexOf("&", start);
+    if (end == -1) {
+      return "";
+    }
+    return encodedSpan.substring(start, end);
+  }
+
+  // class
+  private static final class OpenCensusMessageReceiver implements MessageReceiver {
+    private final MessageReceiver receiver;
+
+    private OpenCensusMessageReceiver(MessageReceiver receiver) {
+      this.receiver = receiver;
+    }
+
+    @Override
+    public void receiveMessage(PubsubMessage message, AckReplyConsumer consumer) {
```

Comment:
Do we need configuration here as well, so that the user can define whether or not they want this?

Reply:
I think that makes sense too.